<!-- # creator: Elizabeth Brooks -->
<!-- # updated: 1 July 2024 -->

---
title: "DA Report"
output: html_document
params:
  analysisIn: NA
  inputDataIn: NA
  targetsIn: NA
  cutLFCIn: NA
  cutFDRIn: NA
  dispersionsIn: NA
  comparisonIn: NA
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.keep='all')
```

# Package Installation & Loading
```{r}
# install any missing packages
packageList <- c("BiocManager", "shiny", "shinythemes", "ggplot2", "rcartocolor", "dplyr", "statmod", "pheatmap", "ggplotify")
biocList <- c("edgeR")
newPackages <- packageList[!(packageList %in% installed.packages()[,"Package"])]
newBioc <- biocList[!(biocList %in% installed.packages()[,"Package"])]
if(length(newPackages)){
  install.packages(newPackages)
}
if(length(newBioc)){
  BiocManager::install(newBioc)
}

# load packages 
suppressPackageStartupMessages({
  library(shiny)
  library(shinythemes)
  library(ggplot2)
  library(rcartocolor)
  library(edgeR)
  library(dplyr)
  library(pheatmap)
  library(ggplotify)
})
```

# Plotting Color Palettes
```{r, fig.keep='all'}
#  plotting palettes
plotColors <- carto_pal(12, "Safe")
plotColorSubset <- c(plotColors[4], plotColors[5], plotColors[6])
```

# Data Input
```{r, fig.keep='all'}
# set analysis type
analysisType <- params$analysisIn

# import gene count data
inputData <- read.csv(file=params$inputDataIn, row.names=1)

# trim the data table of htseq stats
removeList <- c("__no_feature", "__ambiguous", "__too_low_aQual", "__not_aligned", "__alignment_not_unique")
countsTable <- inputData[!row.names(inputData) %in% removeList,]

# import grouping factor
targets <- read.csv(file=params$targetsIn, row.names=1)

# set LFC cut off
cutLFC <- params$cutLFCIn

# set FDR cut off
cutFDR <- params$cutFDRIn

# set dispersions value
dispersionsVal <- params$dispersionsIn

# set comparison
comparisonExp <- params$comparisonIn
```

# Analysis Setup
```{r, fig.keep='all'}
#Setup a design matrix
group <- factor(targets[,1])
#cbind(targets,Group=group)
#Create DGE list object
list <- DGEList(counts=countsTable,group=group)
colnames(list) <- rownames(targets)

#Plot the library sizes before normalization
#jpeg("glmQLF_plotBarsBefore.jpg")
barplot(list$samples$lib.size*1e-6, names=1:ncol(list), ylab="Library size (millions)")
#dev.off()

#Retain genes only if it is expressed at a minimum level
keep <- filterByExpr(list)
summary(keep)
list <- list[keep, , keep.lib.sizes=FALSE]

#Use TMM normalization to eliminate composition biases
# between libraries
list <- calcNormFactors(list)
#list$samples
#Write normalized counts to file
normList <- cpm(list, normalized.lib.sizes=TRUE)
# add gene row name tag
normList <- as_tibble(normList, rownames = "gene")
#write.table(normList, file="glmQLF_normalizedCounts.csv", sep=",", row.names=FALSE, quote=FALSE)

#Write log transformed normalized counts to file
normListLog <- cpm(list, normalized.lib.sizes=TRUE, log=TRUE)
#write.table(normListLog, file="glmQLF_normalizedCounts_logTransformed.csv", sep=",", row.names=TRUE, quote=FALSE)

# calculate the log CPM of the gene count data
logcpm <- cpm(list, log=TRUE)

# combine all columns into one period separated
exp_factor <- data.frame(Sample = unlist(targets, use.names = FALSE))
rownames(exp_factor) <- colnames(logcpm)

# create heatmap for samples
samplesHeat <- as.ggplot(
  pheatmap(logcpm, scale="row", annotation_col = exp_factor, 
           main="Heatmap of RNA-seq Samples", show_rownames = FALSE, fontsize = 12,
           color = colorRampPalette(c(plotColors[5], "white", plotColors[6]))(100))
)

#Verify TMM normalization using a MD plot
#Write plot to file
#jpeg("glmQLF_plotMDBefore.jpg")
plotMD(cpm(list, log=TRUE), column=1)
abline(h=0, col=plotColorSubset[3], lty=2, lwd=2)
#dev.off()

# check the input dispersion value for indication of replication
if(tolower(valueDisp()) == "common" ||
       tolower(valueDisp()) == "trended" ||
       tolower(valueDisp()) == "tagwise" ||
       tolower(valueDisp()) == "auto" ||
       tolower(valueDisp()) == "null"){ # case insensitive check string inputs
  #Use a MDS plot to visualizes the differences
  # between the expression profiles of different samples
  points <- c(0,1,2,3,15,16,17,18)
  colors <- rep(c(plotColors[4], plotColors[5], plotColors[6], plotColors[11]), 2)
  #Write plot with legend to file
  #jpeg("glmQLF_plotMDS.jpg")
  par(mar=c(5.1, 4.1, 4.1, 11.1), xpd=TRUE)
  plotMDS(list, col=colors[group], pch=points[group])
  legend("topright", inset=c(-0.8,0), legend=levels(group), pch=points, col=colors, ncol=2)
  #legend("topleft", legend=levels(group), pch=points, col=colors, ncol=2)
  dev.off()
  #Write plot without legend to file
  #jpeg("glmQLF_plotMDS_noLegend.jpg")
  plotMDS(list, col=colors[group], pch=points[group])
  #dev.off()
  
  # Create a PCA plot with a legend
  #jpeg("glmQLF_plotPCA.jpg")
  par(mar=c(5.1, 4.1, 4.1, 11.1), xpd=TRUE)
  plotMDS(list, col=colors[group], pch=points[group], gene.selection="common")
  legend("topright", inset=c(-0.8,0), legend=levels(group), pch=points, col=colors, ncol=2)
  #legend("topleft", legend=levels(group), pch=points, col=colors, ncol=2)
  dev.off()
  
  # Create a PCA plot without a legend
  #jpeg("glmQLF_plotPCA_noLegend.jpg")
  plotMDS(list, col=colors[group], pch=points[group], gene.selection="common")
  #dev.off()
}
```

# Analysis
```{r, fig.keep='all'}
# check input analysis type
if(analysisType == tolower("pairwise")){
  ##
  # Pairwise Comparisons (Contrasts)
  ##
  
  # check the input dispersion value for indication of replication
  if(tolower(dispersionsVal) == tolower("common") ||
     tolower(dispersionsVal) == tolower("trended") ||
     tolower(dispersionsVal) == tolower("tagwise") ||
     tolower(dispersionsVal) == tolower("auto")){ # case insensitive check string inputs
    # estimate common dispersion and tagwise dispersions to produce a matrix of pseudo-counts
    listDisp <- estimateDisp(list)
    # perform exact test
    tested <- exactTest(listDisp, pair=comparisonExp)
  }else{ # assumes numeric
    # perform exact test
    tested <- exactTest(list, pair=comparisonExp, dispersion=as.numeric(dispersionsVal))
  }
}else if(analysisType == "GLM"){
  ##
  # GLM Fitting
  ##
  
  #The experimental design is parametrized with a one-way layout, 
  # where one coefficient is assigned to each group
  design <- model.matrix(~ 0 + group)
  colnames(design) <- levels(group)
  #design
  
  # check the input dispersion value
  if(tolower(dispersionsVal) == "null"){ # case insensitive check for NULL string input
    #Next, the NB dispersion is estimated
    listDisp <- estimateDisp(list, design, robust=TRUE)
    #listDisp$common.dispersion
    #Visualize the dispersion estimates with a BCV plot
    #Write plot to file
    #jpeg("glmQLF_plotBCV.jpg")
    plotBCV(listDisp)
    #dev.off()
    # estimate the QL dispersions using the data object to estimate dispersion values
    fit <- glmQLFit(listDisp, design, robust=TRUE)
    #head(fit$coefficients)
    #Write plot to file
    #jpeg("glmQLF_plotQLDisp.jpg")
    plotQLDisp(fit)
    #dev.off()
    # view column order
    colnames(fit)
    ##
    # GLM Contrasts
    ##
    # testing explicit nested contrast
    glmContrast <- makeContrasts(glmSet = comparisonExp,
                                 levels = design)
    # results table
    tested <- glmTreat(fit, contrast=glmContrast, lfc=cutLFC)
  }else{ # assumes numeric
    # replace the dispersion value in the DGE object
    list$common.dispersion <- as.numeric(dispersionsVal)
    # estimate the QL dispersions using the input dispersion value(s)
    fit <- glmFit(list, design)
    ##
    # GLM Contrasts
    ##
    # testing explicit nested contrast
    glmContrast <- makeContrasts(glmSet = comparisonExp,
                                 levels = design)
    # results table
    tested <- glmLRT(fit, contrast=glmContrast)
  }
}

# summary table
summary(decideTests(tested))
```

# Results Visualization
```{r, fig.keep='all'}
# create a results table of DE genes by FDR and LFC
resultsTbl <- topTags(tested, n=nrow(tested$table), adjust.method="fdr", p.value=cutFDR)$table
# export tables of DE genes
#Write tags table of DE genes to file
#write.table(resultsTbl, file="glmQLF_topTags.csv", sep=",", row.names=TRUE, quote=FALSE)

# calculate the log2 CPM of the gene count data
logcpm <- cpm(list, log=TRUE)

# identify significantly DE genes
DGESubset <- resultsTbl[resultsTbl$logFC > cutLFC | resultsTbl$logFC < (-1*cutLFC),]
# subset the log2 CPM by the DGE set
DGESubset.keep <- rownames(logcpm) %in% rownames(DGESubset)
logcpmSubset <- logcpm[DGESubset.keep, ]

# combine all columns into one period separated
exp_factor <- data.frame(Sample = unlist(targets, use.names = FALSE))
rownames(exp_factor) <- colnames(logcpmSubset)

# TO-DO: use color blind safe pallette for sample dendrogram
#Create heatmap for DGE
resultsHeat <- as.ggplot(
  pheatmap(logcpmSubset, scale="row", annotation_col = exp_factor, 
           main="Heatmap of Pairwise DE Genes", show_rownames = FALSE,
           color = colorRampPalette(c(plotColors[5], "white", plotColors[6]))(100))
)

# MD plots
#Write plot to file
#jpeg("glmQLF_plotMD.jpg")
# return MD plot
plotMD(tested, main = "Mean-Differences of Pairwise DE Genes", adjust.method="fdr", p.value=cutFDR)
# add blue lines to indicate 2-fold changes
abline(h=c((-1*cutLFC), cutLFC), col="blue") 
#dev.off()

# Volcano plots
# create a results table of DE genes
resultsTbl <- topTags(tested, n=nrow(tested$table), adjust.method="fdr")$table
# add column for identifying direction of DE gene expression
resultsTbl$colorDE <- plotColors[5]
resultsTbl$alphaDE <- 0.75
# identify significantly up DE genes
resultsTbl$colorDE[resultsTbl$logFC > cutLFC & resultsTbl$FDR < cutFDR] <- plotColors[4]
resultsTbl$alphaDE[resultsTbl$logFC > cutLFC & resultsTbl$FDR < cutFDR] <- 1
#resultsTbl$colorDE[sign(resultsTbl$logFC) == 1 & resultsTbl$FDR < cutFDR] <- "Up"
# identify significantly down DE genes
resultsTbl$colorDE[resultsTbl$logFC < (-1*cutLFC) & resultsTbl$FDR < cutFDR] <- plotColors[6]
resultsTbl$alphaDE[resultsTbl$logFC < (-1*cutLFC) & resultsTbl$FDR < cutFDR] <- 1
#resultsTbl$colorDE[sign(resultsTbl$logFC) == -1 & resultsTbl$FDR < cutFDR] <- "Down"
# add column with -log10(FDR) values
resultsTbl$negLog10FDR <- -log10(resultsTbl$FDR)

# create volcano plot
ggplot(data=resultsTbl, aes(x=logFC, y=negLog10FDR, color = colorDE, alpha = alphaDE)) + 
  geom_point() +
  theme_minimal() +
  scale_color_identity() +
  scale_alpha(guide = 'none') +
  ggtitle("Volcano Plot of Pairwise DE Genes") +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(plot.title = element_text(face="bold")) +
  xlab("LFC")
```
